### PR TITLE
[ADD] iot: add iot serial drivers

### PR DESCRIPTION
**Pull request by pha-odoo**
So far there are no IoT drivers for serial equipment.

This adds a base IoT serial driver as well as a base IoT serial scale driver and drivers for the two previously supported scales. For the moment the scale drivers support both the old routes and the new event and action routes.

Task: 1892412

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/point_of_sale/static/src/js/screens.js [hunk 1]
```diff
@@ -314,14 +314,17 @@ var ScaleScreenWidget = ScreenWidget.extend({
...
 
+    _read_scale: function() {
         queue.schedule(function(){
             return self.pos.proxy.scale_read().then(function(weight){
                 self.set_weight(weight.weight);
             });
         },{duration:500, repeat: true});
-
     },
+
```
Comment (qle-odoo) [inline addons/point_of_sale/static/src/js/screens.js]: diff

**Comment by sle-odoo** [comment]
please review the commit messages

File: addons/hw_drivers/controllers/driver.py [hunk 2]
```diff
@@ -1,6 +1,9 @@
 #!/usr/bin/python3
 import logging
 import time
+import inspect
+from abc import ABCMeta
```
Comment (sle-odoo) [inline addons/hw_drivers/controllers/driver.py]: why this change?

File: addons/hw_drivers/controllers/driver.py [hunk 3]
```diff
@@ -1,6 +1,9 @@
...
+from collections import deque
```
Comment (sle-odoo) [inline addons/hw_drivers/controllers/driver.py]: why?

File: addons/hw_drivers/controllers/driver.py [hunk 4]
```diff
@@ -116,22 +120,30 @@ def connect_box(self, token):
...
+class DriverMetaClass(ABCMeta):
     def __new__(cls, clsname, bases, attrs):
         newclass = super(DriverMetaClass, cls).__new__(cls, clsname, bases, attrs)
-        drivers.append(newclass)
+        #  We only add concrete classes to `drivers`.
+        if inspect.isabstract(newclass):
+            return newclass
+        # Some drivers must be tried only when all the others have been ruled out. These are kept at the bottom of the list.
+        if newclass.is_tested_last:
+            drivers.append(newclass)
```
Comment (sle-odoo) [inline addons/hw_drivers/controllers/driver.py]: what if two drivers are `is_tested_last`? wouldn't a sequence be more appropriate?

File: addons/hw_drivers/drivers/SerialBaseDriver.py [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+        'timeout': PROBING_TIMEOUT if is_probing else protocol.timeout,               # longer timeouts for probing
+        'writeTimeout': PROBING_TIMEOUT if is_probing else protocol.writeTimeout      # longer timeouts for probing
+    }
+    connection = serial.Serial(path, **port_config)
+    yield connection
+    connection.close()
+
+
+class DeviceStatus(object):
+    """Represents the current status of a device."""
```
Comment (sle-odoo) [inline addons/hw_drivers/drivers/SerialBaseDriver.py]: is this new class supposed to be used outside of the serial drivers someday?

File: addons/hw_drivers/drivers/SerialScaleDriver.py [hunk 6]
```diff
@@ -0,0 +1,284 @@
...
+
+
+# Ensures compatibility with older versions of Odoo
+class OldScaleDriver(http.Controller):
+    @http.route('/hw_proxy/scale_read', type='json', auth='none', cors='*')
+    def scale_read(self):
+        if ACTIVE_SCALE:
+            ACTIVE_SCALE._scale_read_old_route()
+            return {'weight': ACTIVE_SCALE.data['value']}
+        return None
```
Comment (sle-odoo) [inline addons/hw_drivers/drivers/SerialScaleDriver.py]: do we want the backward compatibility?
also, this kind of things should be clearly explained in the commit message

File: addons/hw_posbox_homepage/controllers/main.py [hunk 7]
```diff
@@ -90,13 +90,15 @@ def get_homepage_data(self):
...
         iot_device = []
-        for status in pos_device:
-            if pos_device[status]['status'] == 'connected':
-                iot_device.append({
-                    'name': status,
-                    'type': 'device',
-                    'message': ' '.join(pos_device[status]['messages'])
-                })
+
+        if not iot_devices:
```
Comment (sle-odoo) [inline addons/hw_posbox_homepage/controllers/main.py]: why this change? why in this commit?

File: addons/point_of_sale/static/src/js/devices.js [hunk 8]
```diff
@@ -427,6 +427,7 @@ var ProxyDevice  = core.Class.extend(mixins.PropertiesMixin,{
         return new Promise(function (resolve, reject) {
             self.message('scale_read',{})
             .then(function (weight) {
+                console.log(weight)
```
Comment (sle-odoo) [inline addons/point_of_sale/static/src/js/devices.js]: to remove i suppose

File: addons/point_of_sale/static/src/js/screens.js [hunk 9]
```diff
@@ -314,7 +313,12 @@ var ScaleScreenWidget = ScreenWidget.extend({
             // add product *after* switching screen to scroll properly
             self.order_product();
         });
+        this._read_scale();
+    },
 
+    _read_scale: function() {
+        var self = this;
+        var queue = this.pos.proxy_queue;
```
Comment (sle-odoo) [inline addons/point_of_sale/static/src/js/screens.js]: i suppose this change is done for an easier override in another addon, i suppose this should be split in another commit and explained

File: addons/hw_drivers/controllers/driver.py (hunk 2 above)
Comment (pha-odoo) [reply addons/hw_drivers/controllers/driver.py]: At first there was some code in the base scale supported method, which raised an exception when it was tested. Adding the abstractmethod decorator to supported and checking for abstract classes to keep base classes out of the drivers list was a way to avoid that. It is not needed anymore but I figured it made sense to keep base classes out of the list, so supported methods can be partially implemented or call not implemented methods. What do you think?

File: addons/hw_drivers/controllers/driver.py (hunk 3 above)
Comment (pha-odoo) [reply addons/hw_drivers/controllers/driver.py]: I'm used to use it when having to append and prepend to a list but I suppose list.insert[0] makes more sense here since there is no performance issue. I'll change it.

File: addons/hw_drivers/controllers/driver.py (hunk 4 above)
Comment (pha-odoo) [reply addons/hw_drivers/controllers/driver.py]: It would, but we don't intend to keep this very long, and there won't be new scale drivers until then. In the near future we'll let the user choose a driver for each port in a wizard.

File: addons/hw_drivers/drivers/SerialBaseDriver.py (hunk 5 above)
Comment (pha-odoo) [reply addons/hw_drivers/drivers/SerialBaseDriver.py]: I doubt it. It seemed clearer somehow to have it separate, but I can integrate the methods in the base serial driver if that's what you mean.

File: addons/hw_posbox_homepage/controllers/main.py (hunk 7 above)
Comment (pha-odoo) [reply addons/hw_posbox_homepage/controllers/main.py]: As a consequence of the changes made for backward compatibility a 'scale' driver appears in the IoT box homepage under device. This is the way to avoid it. It should go on the next IoT box image. Should I put that in the different PR?

File: addons/point_of_sale/static/src/js/devices.js (hunk 8 above)
Comment (pha-odoo) [reply addons/point_of_sale/static/src/js/devices.js]: yes

File: addons/point_of_sale/static/src/js/screens.js (hunk 9 above)
Comment (pha-odoo) [reply addons/point_of_sale/static/src/js/screens.js]: Indeed. It has its own commit with a message explaining the reason for the change.

File: addons/hw_drivers/drivers/SerialScaleDriver.py (hunk 6 above)
Comment (pha-odoo) [reply addons/hw_drivers/drivers/SerialScaleDriver.py]: Yes we really need it. I explained it in the commit message.

File: addons/hw_drivers/controllers/driver.py (hunk 2 above)
Comment (pimodoo) [reply addons/hw_drivers/controllers/driver.py]: @pha-odoo You should just check if 'supported' is implemented to detect if you are in real driver, as implementing it is the only way to have deveice that works.

File: addons/hw_drivers/drivers/SerialScaleDriver.py [hunk 10]
```diff
@@ -0,0 +1,308 @@
...
+    zeroCommand=b'Z',
+    tareCommand=b'T',
+    clearCommand=None,  # No clear command -> Tare again
+    emptyAnswerValid=True,  # AZExtra does not answer unless a new non-zero weight has been detected
+    autoResetWeight=True,  # AZExtra will not return 0 after removing products
+)
+
+
+# Ensures compatibility with older versions of Odoo
+class OldScaleDriver(http.Controller):
```
Comment (qle-odoo) [inline addons/hw_drivers/drivers/SerialScaleDriver.py]: scale_read_old_route

**Comment by qle-odoo** [comment]
@robodoo r+ merge
